### PR TITLE
Reference for stage 3 regex-escaping

### PR DESCRIPTION
**Pull request by Josh-Cena**
Part of https://github.com/mdn/content/issues/36915

### CODE REVIEW

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+{{Specifications}}
+
+## Browser compatibility
+
+{{Compat}}
+
+## See also
+
+- {{jsxref("RegExp")}}
```
Comment (zloirock) [inline files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:95]: Could you add a link to `core-js` polyfill?

https://github.com/zloirock/core-js#regexp-escaping

Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:95]:
```suggestion
- [Polyfill of `RegExp.escape` in `core-js`](https://github.com/zloirock/core-js#regexp-escaping)
- {{jsxref("RegExp")}}
```

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  - : The string to escape.
+
+### Return value
+
+A new string that can be safely used as a literal pattern for the {{jsxref("RegExp/RegExp", "RegExp()")}} constructor. Namely, the following things in the input string are replaced:
+
+- The first character of the string, if it's either a decimal digit (0–9) or ASCII letter (a–z, A–Z), is escaped using the `\x` [character escape](/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_escape) syntax. For example, `RegExp.escape("foo")` returns `"\\x66oo"` (here and after, the two backslashes in a string literal denote a single backslash character). This step ensures that if this escaped string is embedded into a bigger pattern where it's immediately preceded by `\0`, `\c`, `\x0`, `\u000`, etc., the leading character doesn't get interpreted as part of the escape sequence.
+- Regex [syntax characters](/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Literal_character#description), including `^`, `$`, `\`, `.`, `*`, `+`, `?`, `(`, `)`, `[`, `]`, `{`, `}`, and `|`, as well as the `/` delimiter, are escaped by inserting a `\` character before them. For example, `RegExp.escape("foo.bar")` returns `"\\x66oo\\.bar"`.
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:30]: >  For example, `RegExp.escape("foo.bar")` returns `"\\x66oo\\.bar"`.

An example that escapes more than one syntax character?

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  - : Thrown if `string` is not a string.
+
+## Examples
+
+### Using RegExp.escape()
+
+The following examples demonstrate various inputs and outputs for the `RegExp.escape()` method.
+
+```js
+RegExp.escape("foo.bar"); // "\\x66oo\\.bar"
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:49]: I quite like something with spaces - like the very simple examples from the proposal (just one, your call).
```
RegExp.escape("Buy it. use it. break it. fix it.") 
RegExp.escape("The Quick Brown Fox");
```

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+const input =
+  "Considering using [RegExp.escape()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/escape) to escape special characters in a string.";
+const domain = "developer.mozilla.org";
+console.log(removeDomain(input, domain));
+// Considering using [RegExp.escape()](/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/escape) to escape special characters in a string.
+```
+
+The above inadvertently converts the literal "." character to the regex [wildcard](/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Wildcard) character. This means it will match anything, including `developer-mozilla.org`, which is not what we intended. To fix this, we can use `RegExp.escape()` to ensure that any user input is treated as a literal pattern:
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:72]: It may seem pedantic, but you don't have  `developer-mozilla.org` in the text you're matching, so it feels a bit "is this a real problem-ish". Perhaps consider

```suggestion
Inserting the `domain` above results in the regular expression literal `https?://developer.mozilla.org(?=/)`, where the "." character is a regex [wildcard](/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Wildcard) character. This means the string will match the string with any character in place of the ".", such as `developer-mozilla-org`.
While we have no problematic strings in the particular input shown here, this is not robust code.

To fix this, we can use `RegExp.escape()` to ensure that any user input is treated as a literal pattern:
```

**Comment by hamishwillee** [comment]
@Josh-Cena Looks great. I don't fully understand [this gist linked from the proposal](https://gist.github.com/bakkot/5a22c8c13ce269f6da46c7f7e56d3c3f). But I take it to mean there might be risky places to insert the literal, even escaped. Is this any kind of concern, or should I ignore this? I assume it isn't something you consider worth mentioning - just being careful.

As an aside, you link to the escape sequence https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions#escape_sequences
The proposal escapes space in its example - I don't see space mentioned in the list. Might not be important - just noting in case it is an omission.

File: files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md [hunk 5]
```diff
@@ -41,7 +41,7 @@ A new string, with all matches of a pattern replaced by a replacement.
 
 This method does not mutate the string value it's called on. It returns a new string.
 
-Unlike [`replace()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace), this method would replace all occurrences of a string, not just the first one. This is especially useful if the string is not statically known, as calling the [`RegExp()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/RegExp) constructor without escaping special characters may unintentionally change its semantics.
+Unlike [`replace()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace), this method would replace all occurrences of a string, not just the first one. This is especially useful if the string is not statically known, as calling the [`RegExp()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/RegExp) constructor without escaping special characters may unintentionally change its semantics. (You can also use {{jsxref("RegExp.escape()")}} to make the replacement string a literal pattern, but that is more indirection than just calling `replaceAll()`.)
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md:44]: I'm not sure what you're trying to say with ", but that is more indirection than just calling `replaceAll()`.)"

It reads as though you are not recommending using `RegExp.escape()` for this case of the string for the regexp constructor.

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 2 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:30]:
```suggestion
- Regex [syntax characters](/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Literal_character#description), including `^`, `$`, `\`, `.`, `*`, `+`, `?`, `(`, `)`, `[`, `]`, `{`, `}`, and `|`, as well as the `/` delimiter, are escaped by inserting a `\` character before them. For example, `RegExp.escape("foo.bar")` returns `"\\x66oo\\.bar"`, and `RegExp.escape("(foo)")` returns `"\\(foo\\)"`.
```

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 3 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:49]:
```suggestion
RegExp.escape("Buy it. use it. break it. fix it.");
// "Buy\\ it\\.\\ use\\ it\\.\\ break\\ it\\.\\ fix\\ it\\."
RegExp.escape("foo.bar"); // "\\x66oo\\.bar"
```

File: files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md (hunk 5 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md:44]: Aye, I'm not recommending `RegExp.escape`. I'm just preempting someone saying, "Buuuut you can actually use `replace() + RegExp()`, you just need to escape it!!"

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 3 above)
Comment (hamishwillee) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:51]: @Josh-Cena Shouldn't the space be represented by `\\x20`?

File: files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md (hunk 5 above)
Comment (hamishwillee) [reply files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md:44]: Perhaps it is actually the "This is especially useful if" part that is confusing me - why is being able to replace all occurrences of a string especially useful if the string is not statically known?

Or to put it another way, don't I have to escape the pattern I use in this method too?

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 3 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:51]: Oh you right.

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+## Examples
+
+### Using RegExp.escape()
+
+The following examples demonstrate various inputs and outputs for the `RegExp.escape()` method.
+
+```js
+RegExp.escape("Buy it. use it. break it. fix it.");
+// "Buy\\ it\\.\\ use\\ it\\.\\ break\\ it\\.\\ fix\\ it\\."
```
Comment (Josh-Cena) [inline files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:50]:
```suggestion
// "Buy\\x20it\\.\\x20use\\x20it\\.\\x20break\\x20it\\.\\x20fix\\x20it\\."
```

File: files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md (hunk 5 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md:44]: > why is being able to replace all occurrences of a string especially useful if the string is not statically known?

I mean, that doesn't sound like an obscure use case at all, is it? Maybe the replacement string comes from user configuration, maybe it's retrieved from a database...

> don't I have to escape the pattern I use in this method too?

In `replaceAll`? No, since the idea is that you can just do `replaceAll(oldString, newString)`. The problem with `replace(oldString, newString)` is that it only does the replacement once, so if you want to replace all occurrences of `oldString`, you have to first convert it to a global regex, but the conversion is non-trivial.

Comment (hamishwillee) [reply files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md:44]: I understand why you wouldn't use `replace()` and I understand why you would use `replaceAll` for a static string.

But `replaceAll` takes a pattern that can be a regexp

Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md:44]: I assume by "static string" you just mean "string"? Since it doesn't matter where the string comes from. For regexp there's a difference between "static regexp" and "dynamically assembled regexp" though since the latter is subject to regex syntax injection. String replacement is safe. Regex replacement should preferably only be used with static regexes. It happens that `replace()` and `replaceAll()`, when taking a regex, have exactly the same behavior (and `replaceAll()` throws if the regex is non-global), so yeah you could say that using `replaceAll` with a regex is useless since it's always equivalent to a `replace` anyway.

**Comment by Josh-Cena** [comment]
Oh I forgot to ask but @ljharb @bakkot your reviews will be appreciated too. Thanks!

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+## Examples
+
+### Using RegExp.escape()
+
+The following examples demonstrate various inputs and outputs for the `RegExp.escape()` method.
+
+```js
+RegExp.escape("Buy it. use it. break it. fix it.");
+// "Buy\\x20it\\.\\x20use\\x20it\\.\\x20break\\x20it\\.\\x20fix\\x20it\\."
```
Comment (bakkot) [inline files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:50]:
```suggestion
// "\\x42uy\\x20it\\.\\x20use\\x20it\\.\\x20break\\x20it\\.\\x20fix\\x20it\\."
```
Leading ASCII letters are escaped as well, as of https://github.com/tc39/proposal-regex-escaping/pull/77.

I assume this example is from the readme; my bad for failing to update it when I updated the spec.

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+slug: Web/JavaScript/Reference/Global_Objects/RegExp/escape
+page-type: javascript-static-method
+browser-compat: javascript.builtins.RegExp.escape
+---
+
+{{JSRef}}
+
+The **`RegExp.escape()`** static method [escapes](/en-US/docs/Web/JavaScript/Reference/Regular_expressions#escape_sequences) any potential regex syntax characters in a string, and returns a new string that can be safely used as a [literal](/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Literal_character) pattern for the {{jsxref("RegExp/RegExp", "RegExp()")}} constructor.
+
+Always prefer this function to a manual search & replace using, for example, {{jsxref("String.prototype.replaceAll()")}}, because `RegExp.escape()` can handle more edge cases and use the right escape sequence that doesn't cause syntax errors in certain contexts.
```
Comment (ljharb) [inline files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:12]: i'm not sure about this note? it's certainly better than using replaceAll with `new RegExp`, but not better than using it with a string, I think.

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+const input =
+  "Considering using [RegExp.escape()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/escape) to escape special characters in a string.";
+const domain = "developer.mozilla.org";
+console.log(removeDomain(input, domain));
+// Considering using [RegExp.escape()](/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/escape) to escape special characters in a string.
+```
+
+Inserting the `domain` above results in the regular expression literal `https?://developer.mozilla.org(?=/)`, where the "." character is a regex [wildcard](/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Wildcard) character. This means the string will match the string with any character in place of the ".", such as `developer-mozilla-org`.
+While we have no problematic strings in the particular input shown here, this is not robust code.
```
Comment (ljharb) [inline files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:75]: should the input _be_ problematic, then?

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+{{Specifications}}
+
+## Browser compatibility
+
+{{Compat}}
+
+## See also
+
+- [Polyfill of `RegExp.escape` in `core-js`](https://github.com/zloirock/core-js#regexp-escaping)
```
Comment (ljharb) [inline files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:100]: are you still only linking to core-js polyfills? if not, it'd be great to include https://www.npmjs.com/package/regexp.escape

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 8 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:12]: I mean you should use `RegExp.escape(string)` instead of `string.replaceAll(/[()[\]{}^$...]/g, "\\$&")`, which was the alternative before.

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 10 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:107]: I personally would not object, but I would need to see what others say. In the meantime let's hold the same policy that other maintainers have found acceptable, which is to only consistently include core-js polyfills. Hope you would understand.

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 8 above)
Comment (ljharb) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:12]: Is there something about the paragraph that implies that the `replaceAll` argument is a regex? replaceAll also takes a string, so the ideal would be to just provide it a string directly, no?

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 10 above)
Comment (ljharb) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:107]: Sure. Where can I go to revisit this discussion? Last time I tried it never went anywhere, and another polyfill maintainer tried recently and it seemed to get overly heated.

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 8 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:12]: Oh, I meant "you should use `RegExp.escape` for the use case of escaping of regex syntax, instead of implementing it yourself using `replaceAll`". It has nothing to do with "escaping a regex so it can be passed to `replaceAll`".

Comment (ljharb) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:12]: How about, always prefer this function when dynamically creating a RegExp with user-provided content? replace all really doesn’t have anything to do with it imo.

Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:12]: How about this?

```suggestion
When dynamically creating a {{jsxref("RegExp")}} with user-provided content, always consider using this function to sanitize the input, unless the input is actually intended to contain regex syntax. In addition, don't try to re-implement its functionality by, for example, using {{jsxref("String.prototype.replaceAll()")}} to insert a `\` before all syntax characters, because `RegExp.escape()` can handle more edge cases and use the right escape sequence that doesn't cause syntax errors in certain contexts.
```

Comment (ljharb) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:12]: Much better, thank you!

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+### Parameters
+
+- `string`
+  - : The string to escape.
+
+### Return value
+
+A new string that can be safely used as a literal pattern for the {{jsxref("RegExp/RegExp", "RegExp()")}} constructor. Namely, the following things in the input string are replaced:
+
+- The first character of the string, if it's either a decimal digit (0–9) or ASCII letter (a–z, A–Z), is escaped using the `\x` [character escape](/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_escape) syntax. For example, `RegExp.escape("foo")` returns `"\\x66oo"` (here and after, the two backslashes in a string literal denote a single backslash character). This step ensures that if this escaped string is embedded into a bigger pattern where it's immediately preceded by `\0`, `\c`, `\x0`, `\u000`, etc., the leading character doesn't get interpreted as part of the escape sequence.
```
Comment (bakkot) [inline files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:29]:
```suggestion
- The first character of the string, if it's either a decimal digit (0–9) or ASCII letter (a–z, A–Z), is escaped using the `\x` [character escape](/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_escape) syntax. For example, `RegExp.escape("foo")` returns `"\\x66oo"` (here and after, the two backslashes in a string literal denote a single backslash character). This step ensures that if this escaped string is embedded into a bigger pattern where it's immediately preceded by `\1`, `\x0`, `\u000`, etc., the leading character doesn't get interpreted as part of the escape sequence.
```
It's true that the escapes are also necessary for the contexts after `\0` and `\c`, but understanding why requires knowledge of the arcana of the regexp grammar which no one should have. Whereas `\1` is something readers actually should be familiar with.

In other words, I don't want people to read this sentence and discover that `\cA` or `\06` are things they could put in their regexes, even though it is technically accurate that these contexts are among those we considered when designing this feature.

Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:29]: Wait, is `\cA` deprecated? It's still mentioned in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_escape. Also for `\0`, my intention was that "this `\0` was actually supposed to be a null character by itself, but now it's inadvertently joined with something else". OTOH, `\1` is mentioned in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions#escape_sequences as already-deprecated:

> `\` followed by any other digit character becomes a [legacy octal escape sequence](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Deprecated_and_obsolete_features#escape_sequences), which is forbidden in [Unicode-aware mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/unicode#unicode-aware_mode).

So my hope is that `\1` would never appear as a valid regex sequence anyway.

Comment (bakkot) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:29]: > Wait, is `\cA` deprecated?

Not in the sense of being in annex B, but it's still not something people should use (because no one will know what it means).

> now it's inadvertently joined with something else

Yeah but the only way it can be inadvertently joined with something else is to form a legacy octal escape sequence, which people usually shouldn't have to know about.

> So my hope is that `\1` would never appear as a valid regex sequence anyway.

That wording could use updating. `\1` is a perfectly normal thing to appear in a regex as a reference to the first capturing group: `/(.)\1/` is a regex which matches any character appearing twice in a row. Similarly `\12` in a regex with 12 capturing groups. So you don't need to know about legacy octal escape sequences to understand why you need to escape leading digits.

It only falls back to being a legacy octal escape in the case that there are fewer capturing groups than the number it would represent.

Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:29]: Oh. I get what you mean. Thanks!

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 10 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:107]: It looks we still hold https://github.com/orgs/mdn/discussions/475 as our conclusion; i.e. as long as someone steps out to endorse es-shims we are happy to add it, and we should formally document a list of trusted polyfill sources and reject everything not in this list (with another process to endorse more). I'm happy to add es-shims, but I'm too busy to initiate the process in the next few months.

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 8 above)
Comment (hamishwillee) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md]: I like this.

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md [hunk 12]
```diff
@@ -9,7 +9,7 @@ browser-compat: javascript.builtins.RegExp.escape
 
 The **`RegExp.escape()`** static method [escapes](/en-US/docs/Web/JavaScript/Reference/Regular_expressions#escape_sequences) any potential regex syntax characters in a string, and returns a new string that can be safely used as a [literal](/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Literal_character) pattern for the {{jsxref("RegExp/RegExp", "RegExp()")}} constructor.
 
-Always prefer this function to a manual search & replace using, for example, {{jsxref("String.prototype.replaceAll()")}}, because `RegExp.escape()` can handle more edge cases and use the right escape sequence that doesn't cause syntax errors in certain contexts.
+When dynamically creating a {{jsxref("RegExp")}} with user-provided content, always consider using this function to sanitize the input, unless the input is actually intended to contain regex syntax. In addition, don't try to re-implement its functionality by, for example, using {{jsxref("String.prototype.replaceAll()")}} to insert a `\` before all syntax characters, because `RegExp.escape()` can handle more edge cases and use the right escape sequence that doesn't cause syntax errors in certain contexts.
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:12]: Subedit to taste.
```suggestion
When dynamically creating a {{jsxref("RegExp")}} with user-provided content, consider using this function to sanitize the input (unless the input is actually intended to contain regex syntax). In particular, don't try to re-implement its functionality by, for example, using {{jsxref("String.prototype.replaceAll()")}} to insert a `\` before all syntax characters.  `RegExp.escape()` has been designed to use escape sequence that don't cause syntax errors in many more edge cases/contexts than hand-crafted code is likely to achieve.
```

Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:12]: This paragraph is calling out two anti-patterns to avoid:

1. `something.replaceAll(new RegExp(userString), ...)` - use `something.replaceAll(new RegExp(RegExp.escape(userString)), ...)` instead.
2. `userString.replaceAll(/[...]/, "\\$&")` - use `RegExp.escape(userString)` instead.

They are two separate use cases, so I think we should say "in addition" instead of "in particular".

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 10 above)
Comment (ljharb) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:107]: Who would need to endorse it? Who endorsed core-js originally? Does it require an MDN team member to initiate the process, or can I?

Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:107]: > Who endorsed core-js originally?

I think that endorsement was implicitly given due to its wide usage via Babel. Happy to formalize the decision in the meta docs.

> Who would need to endorse it? Does it require an MDN team member to initiate the process, or can I?

Just someone who raised a PR and got it merged without any objections. We don't know exactly yet as it's entirely new to us. I bet you can, too, as long as you get multiple feedback on your PR. I imagine it would be put under either https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/What_we_write/Criteria_for_inclusion or https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Writing_style_guide#external_links, as a new section called "Inclusion of polyfills" that basically restates the conclusion in https://github.com/orgs/mdn/discussions/475 and say "we only allow polyfills from trusted sources, which are the following:"

File: files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md (hunk 5 above)
Comment (hamishwillee) [reply files/en-us/web/javascript/reference/global_objects/string/replaceall/index.md:44]: I get it now.

You are saying that you can also replace all strings using `replace()` with a regexp. However this is not safe if the regexp takes a "user-passed string" unless the passed string is sanitized. You could sanitize it first using the `escape()` method, but if you're doing literal string replacement you might as well just use `replaceAll` with a (non-regexp) string.

The reason this text is hard to parse is that when you read this it isn't obvious that you're saying can also use `replace()` with regexp to replace the strings. Might be worth spelling it out for people like me.

If so, maybe:

```suggestion
Unlike [`replace()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace), this method replaces all occurrences of a string, not just the first one.

Note that it is also possible to use `replace()` with a [`RegExp()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/RegExp) to replace all instances of a string, this can have unintended consequences if the string contains special characters that have meaning in regular expressions (which might happen if the replacement string comes from user input).
While you can mitigate this case using {{jsxref("RegExp.escape()")}} to make the regular expression string into a literal pattern, it is better just to use `replaceAll()`.
```

File: files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md (hunk 12 above)
Comment (hamishwillee) [reply files/en-us/web/javascript/reference/global_objects/regexp/escape/index.md:12]: @Josh-Cena You're right. I've reverted that bit above.

**Comment by hamishwillee** [comment]
This is good to go from my perspective. I'll approve this when you let me know the discussions with the technical experts have wound down.

**Comment by Josh-Cena** [comment]
I believe I've addressed all reviews.

**Review by hamishwillee — APPROVED** [review]

**Comment by hamishwillee** [comment]
Thanks very much for this @Josh-Cena - also @bakkot @ljharb
